cpp/src/parquet/column_reader.h

```diff
@@ -118,6 +120,110 @@ struct CryptoContext {
   std::shared_ptr<Decryptor> data_decryptor;
 };
 
+struct PageSkipInfo {
+  PageSkipInfo() = default;
+
+  const RowRanges::Range& Range() { return ranges_[index_]; }
+
+  int64_t SkipRowNum() const { return skip_row_nums_[index_]; }
+
+  int64_t LastRowIndex() const { return last_row_indices_[index_]; }
+
+  bool HasNext() const {
+    if (index_ != ranges_.size() - 1) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  ::arrow::Status Next() {
+    ARROW_RETURN_IF(!HasNext(), ::arrow::Status::Invalid("No more range to read"));
+    index_++;
+    return ::arrow::Status::OK();
+  }
+
+  int64_t EndRowIndex() const { return last_row_indices_.back(); }
+
+  bool operator==(const PageSkipInfo& other) const {
+    if (ranges_ != other.ranges_) {
+      return false;
+    }
+    if (last_row_indices_ != other.last_row_indices_) {
+      return false;
+    }
+    if (skip_row_nums_ != other.skip_row_nums_) {
+      return false;
+    }
+
+    return true;
+  }
+
+  bool operator!=(const PageSkipInfo& other) const { return !(*this == other); }
+
+  //
+  //   | <--------------------------- column chunk -------------------------------> |
+  //             | <-------------------- page N -----------------------> |
+  //        first_row_idx                                          last_row_idx
+  //   |-- ... --|-------------------------------------------------------|--- ... ---|
+  //                       |---- range0 ----|         |---- range1 ----|
+  //             |--skip0--|                |--skip1--|
+  //             |------last_row_index0-----|
+  //             |-------------------last_row_index1-------------------|
+  //
+
+  // Row ranges for this page, start counting from within column chunk
+  std::vector<RowRanges::Range> ranges_;
+
+  // The num of rows to skip before reading each row range
+  std::vector<int64_t> skip_row_nums_;
+
+  // The last row index for echo row range, start counting from within the page
```

Comment:
What does "echo row range" mean, is it "each row range"?

Reply:
> What does "echo row range" mean, is it "each row range"?

Yes.